### ISSUE TITLE
Documentation on background requests using the threadpool?

### ISSUE THREAD

**Issue by andrewvc** [cheald/manticore#41]
I've looked through the source, but its kind of hard to follow here. Do background requests use the threadpool or do they just create a new thread/future for the background request?

If I have a pool_max of 10, and I issue 11 background requests what happens?

**Comment by cheald** [comment]
Sorry, this is a little obscured due to the use of the Java executor. The executor pool is a `newCachedThreadPool()`, which will grow as large as is needed to service outstanding requests. However, the executor pool is really just there to move execution of the request off of the main thread. The `pool_max` is the maximum number of concurrent requests that the client itself will run - the executor will run all the outstanding requests, which will check out connections from the pool first-come first-served. Requests which cannot check out a connection will block (in their own thread) until one becomes available.

The net effect is that you should only ever see 10 concurrent requests, though any number of threads may be created to service the outstanding requests in the pool. An ideal compromise would be a cached thread pool which grows to a maximum of `pool_max` connections and backs the remaining requests with an unbounded queue, but I don't think that [Executor](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Executors.html) provides that by default.

**Comment by andrewvc** [comment]
Thanks @cheald Got it, and to be clear, there is no blocking behavior on the 11th request yes?

**Comment by andrewvc** [comment]
@cheald I should clarify, if I WANT to essentially have blocking on that 11th request ( I don't want to submit to the pool if it's full) I'd have to manage connections myself correct?

**Comment by cheald** [comment]
Correct; it should accept requests indefinitely. We could work around this by allowing the user to provide a custom ThreadPool, which would allow you to customize the growth and rejection policy - using a bounded queue would result in submits blocking if it's full, or you could use `ThreadPoolExecutor.AbortPolicy` which would throw an exception if you attempt to submit to a full executor pool.

Can you elaborate on the use case? I'd like to think about how to expose an interface for this if it might be a use case that others would use.

**Comment by andrewvc** [comment]
@cheald no worries, I wound up implementing my own blocking behavior using tokens and a `SizedQueue`.

The used case in logstash is that we want backpressure. Basically, we may get one request at a time, but we never want to have more than 10 inflight. If we get the 11th we want to block so as not to keep consuming from our input source. 

This isn't a feature we need that much from Manticore because very soon all of our processing will take place in batches and `#parallel` and `#execute!` perfectly work for that :)

**Comment by cheald** [comment]
Cool. I think I'm going to go ahead and expose an `:executor` option in the Client anyhow, so that if you wanted to provide your own executor (ie, backed with `SizedQueue` if you wanted) you could inject that easily. No downside, and provides flexibility with sane defaults.

**Comment by andrewvc** [comment]
Sounds awesome @cheald !

**Comment by cheald** [comment]
Okay, so the means to do this is already sort of provided, though accidentally. `Client` exposes `#executor`, which you can then modify:

```ruby
@client.executor.set_maximum_pool_size max_pool_size
@client.executor.set_rejected_execution_handler ThreadPoolExecutor::CallerRunsPolicy.new
```

This will cause the executor to create up to `max_pool_size` threads, and then any further requests submitted will instead be executed on the calling thread. This is slightly suboptimal in that if the calling thread's request takes much longer than the backgrounded requests, you end up with unutilized idle time. However, it's generally not considered a good idea to utilize the blocking mechanism in the `ThreadPoolExecutor` [due to potential deadlock issues](http://stackoverflow.com/a/3518588/271475).

So, if you don't want the calling thread running requests, then using a `SizedQueue` in front of the background request queueing is probably the right way to do it.
